### PR TITLE
PLFM 2114, 2124

### PR DESCRIPTION
**Pull request by kaysoky**
- Changes DBOUserGroup to include an extra ETAG column.
- GroupMembers DAO Declaration
- Crowd GroupMembers DAO implementation
- DBO GroupMembers DAO implementation
- Tests (with migration test)
- Worker for migrating Crowd groups to RDS groups

### CODE REVIEW

**Comment by kaysoky** [comment]
These changes are mostly additive and should not break existing tests.  
The possibly breaking change of the UserManager (https://sagebionetworks.jira.com/browse/PLFM-2128) will be included in a separate pull once this one is reviewed/pulled.

**Comment by kaysoky** [comment]
https://github.com/kaysoky/Synapse-Repository-Services/compare/PLFM-2114...PLFM-2128

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+		}
+		
+		try {
+			// This isn't transactional... which is one reason we're migrating away from this
+			for (UserGroup ug : dto.getMembers()) {
+				CrowdAuthUtil.addUserToGroup(group.getName(), ug.getName());
+			}
+			
+			// Update the etags of all affected items
+			List<String> ids = new ArrayList<String>();
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: this dao cannot touch the DB

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	public GroupMembers getMembers(String principalId)
+			throws DatastoreException {
+		
+		GroupMembers members = new GroupMembers();
+		members.setId(principalId);
+		UserGroup group;
+		try {
+			group = userGroupDAO.get(principalId);
+		} catch (NotFoundException e) {
+			throw new DatastoreException(e);
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: this should throw a not found

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+			members.getMembers().addAll(userGroupDAO.getGroupsByNames(usernames).values());
+		}
+		return members;
+	}
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public void addMembers(GroupMembers dto) throws DatastoreException {
+		// Make sure the DTO holds a group, not an individual
+		UserGroup group;
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: throw exception if more than on as there is no transaction.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+				CrowdAuthUtil.removeUserFromGroup(group.getName(), ug.getName());
+			}
+			
+			// Update the etags of all affected items
+			List<String> ids = new ArrayList<String>();
+			ids.add(group.getId());
+			for (UserGroup removed : dto.getMembers()) {
+				ids.add(removed.getId());
+			}
+			userGroupDAO.touchList(ids);
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: again no db touch here.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+			throw new DatastoreException("500 Server Error - "+e.getMessage(), e);
+		} catch (IOException e) {
+			throw new DatastoreException("500 Server Error - "+e.getMessage(), e);
+		}
+	}
+
+	@Override
+	public void removeMembers(GroupMembers dto) throws DatastoreException {
+		UserGroup group;
+		try {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: only one as well

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+			group = userGroupDAO.get(dto.getId());
+		} catch (NotFoundException e) {
+			throw new DatastoreException(e);
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: not found

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+		} catch (AuthenticationException e) {
+			throw new DatastoreException("500 Server Error - "+e.getMessage(), e);
+		} catch (IOException e) {
+			throw new DatastoreException("500 Server Error - "+e.getMessage(), e);
+		}
+	}
+
+	@Override
+	public void removeMembers(GroupMembers dto) throws DatastoreException {
+		UserGroup group;
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: removeMember(string groupId, List<String> listOfIdsToRemove)

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+		if (usernames.size() > 0) {
+			members.getMembers().addAll(userGroupDAO.getGroupsByNames(usernames).values());
+		}
+		return members;
+	}
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public void addMembers(GroupMembers dto) throws DatastoreException {
+		// Make sure the DTO holds a group, not an individual
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: same signature as remove()

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+ *
+ */
+public class CrowdGroupMembersDAO implements GroupMembersDAO {
+	
+	@Autowired
+	private UserGroupDAO userGroupDAO;
+	
+	@Override
+	public GroupMembers getMembers(String principalId)
+			throws DatastoreException {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java]: return List<UserGroup>

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+	private static final String PRINCIPAL_ID_PARAM_NAME = "principalId";
+	private static final String GROUP_ID_PARAM_NAME     = "groupId";
+	private static final String MEMBER_ID_PARAM_NAME    = "memberId";
+	
+	private static final String SELECT_MEMBERS_OF_GROUP = 
+			"SELECT * FROM "+SqlConstants.TABLE_USER_GROUP+
+			" WHERE "+SqlConstants.COL_USER_GROUP_ID+" IN"+
+					" (SELECT "+SqlConstants.COL_GROUP_MEMBERS_MEMBER_ID+" FROM "+SqlConstants.TABLE_GROUP_MEMBERS+
+					" WHERE "+SqlConstants.COL_GROUP_MEMBERS_GROUP_ID+"=:"+PRINCIPAL_ID_PARAM_NAME+")";
+
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java:52]: re-write as a single non-nested

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+		MapSqlParameterSource param = new MapSqlParameterSource();
+		param.addValue(PRINCIPAL_ID_PARAM_NAME, principalId);
+		List<DBOUserGroup> dbos = simpleJdbcTemplate.query(SELECT_MEMBERS_OF_GROUP, userGroupRowMapper, param);
+		
+		for (DBOUserGroup dbo : dbos) {
+			UserGroup dto = new UserGroup();
+			UserGroupUtils.copyDboToDto(dbo, dto);
+			members.getMembers().add(dto);
+		}
+
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java]: update the util to take a list (any other place that would use that should be refactored)

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java [hunk 12]
```diff
@@ -359,4 +398,27 @@ public void bootstrapUsers() throws Exception {
 			}
 		}
 	}
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public void touchList(List<String> ids) {
+		MapSqlParameterSource params[] = new MapSqlParameterSource[ids.size()];
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java]: needs deadlock prevention

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+		return userGroupDAO.get(parentGroupIds);
+	}
+	
+	/**
+	 * Iteratively moves up the group hierarchy and fetches all parents of the given Principal
+	 */
+	private Set<String> getParentGroupIds(String principalId) {
+		Set<String> groupIds = new HashSet<String>();
+		List<String> memberIds = new ArrayList<String>();
+		memberIds.add(principalId);
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java]: rename explored and frontier

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java [hunk 14]
```diff
@@ -359,4 +398,27 @@ public void bootstrapUsers() throws Exception {
...
+		for (int i = 0; i < params.length; i++) {
+			params[i] = new MapSqlParameterSource();
+			params[i].addValue(ID_PARAM_NAME, ids.get(i));
+			params[i].addValue(ETAG_PARAM_NAME, UUID.randomUUID().toString());
+		}
+		simpleJdbcTemplate.batchUpdate(UPDATE_ETAG_LIST, params);
+	}
+
+	@Override
+	public boolean hasChanged(UserGroup dto) {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java]: remove if can

File: services/workers/src/main/resources/grouper-spb.xml [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+		<property name="jobDetail">
+			<bean class="org.springframework.scheduling.quartz.MethodInvokingJobDetailFactoryBean">
+				<property name="targetObject" ref="crowdGroupSynchronizerSemaphoreGatedRunner" />
+				<property name="targetMethod" value="attemptToRun" />
+			</bean>
+		</property>
+		<!-- We stagger the start delay of each trigger to spread out the timing -->
+		<property name="startDelay" value="567" />
+		<property name="repeatInterval" value="300000" />
+	</bean>
```
Comment (john-hill) [inline services/workers/src/main/resources/grouper-spb.xml:41]: make sure this is the same as the current cache time.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+			
+		} catch (NotFoundException e) {
+			throw new DatastoreException("500 Server Error - "+e.getMessage(), e);
+		} catch (IOException e) {
+			throw new DatastoreException("500 Server Error - "+e.getMessage(), e);
+		}
+	}
+
+	@Override
+	public Set<String> markAsUpdated(String groupId, List<String> memberIds)
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:147]: try to make private.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+			params[i].addValue(GROUP_ID_PARAM_NAME, groupId);
+			params[i].addValue(MEMBER_ID_PARAM_NAME, memberId);
+		}
+		simpleJdbcTemplate.batchUpdate(DELETE_MEMBERS_OF_GROUP, params);
+		
+		// Mark each of the children as modified
+		markAsUpdated(groupId, memberIds);
+	}
+	
+	public static byte[] zipParents(List<String> parents) throws IOException {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java:215]: move to utility

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+		param = new MapSqlParameterSource();
+		param.addValue(GROUP_ID_PARAM_NAME, principalId);
+		param.addValue(PARENT_BLOB_PARAM_NAME, cache);
+		simpleJdbcTemplate.update(INSERT_BLOB_INTO_PARENTS_CACHE, param);
+		
+		return userGroupDAO.get(parents);
+	}
+
+	@Override
+	public Set<String> markAsUpdated(String groupId, List<String> memberIds)
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java:290]: multi-thread test.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java [hunk 19]
```diff
@@ -82,15 +86,20 @@
...
 
 	// the pattern is: select x.id, y.etag from g x LEFT OUTER JOIN p y on x.id=y.owner order by x.id
 	// the query is: select g.id, p.etag from user_group g LEFT OUTER JOIN user_profile p on g.id=p.owner_id order by g.id limit l offset o 
-	private static final String SELECT_ALL_PAGINATED_WITH_ETAG = 
-		"SELECT g."+COL_USER_GROUP_ID+", p."+COL_USER_PROFILE_ETAG+" FROM "+
-		TABLE_USER_GROUP+" g LEFT OUTER JOIN "+TABLE_USER_PROFILE+
-		" p ON g."+COL_USER_GROUP_ID+" = p."+COL_USER_PROFILE_ID+
-		" ORDER BY g."+COL_USER_GROUP_ID+" LIMIT :"+LIMIT_PARAM_NAME+
-		" OFFSET :"+OFFSET_PARAM_NAME;
+	// private static final String SELECT_ALL_PAGINATED_WITH_ETAG =
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java:97]: remove

**Pull request merged**